File: ios/Sources/KlinikAuditFeature/AuditModel.swift
import Foundation
import KlinikAPI
import KlinikCore

public enum AuditPhase: Sendable, Equatable {
    case loading
    case loaded
    case empty
    case notPermitted
    case failed(String)
}

public struct AuditState: Sendable, Equatable {
    public var phase: AuditPhase = .loading
    public var entries: [AuditEntry] = []
    public var anomalies: [AuditAnomaly] = []
    public var filter = AuditFilter()
    public var nextCursor: String?
    public var loadingMore = false

    public init() {}

    public var hasMore: Bool { nextCursor != nil }
}

/**
 * The audit trail, for the doctor who owns the clinic (spec M13).
 *
 * Two halves. The trail itself, filtered, because "who opened this file" is a
 * question with an answer. And the anomalies the server detects — a nurse
 * opening two hundred files in a night, access outside working hours, repeated
 * failed sign-ins — which is the half nobody would find by scrolling.
 *
 * Anomalies are shown first and their sentences are the server's own. It knows
 * what it counted and over what window; re-wording that here would be a second
 * description of a thing this client cannot see.
 */
@MainActor
public final class AuditModel {
    private let api: AuditAPI
    private var state = AuditState()

    public init(api: AuditAPI) {
        self.api = api
    }

    public func currentState() -> AuditState { state }

    public func load() async {
        async let entries = attempt { try await self.api.entries(self.state.filter) }
        async let anomalies = attempt { try await self.api.anomalies() }

        let loaded = await (entries, anomalies)

        guard let page = loaded.0.value else {
            // A refusal only when the server refused. An audit log that cannot
            // be reached is not an audit log somebody is barred from.
            switch ReadOutcome.of([loaded.0.error].compactMap { $0 }) {
            case .refused: state.phase = .notPermitted
            case .failed(let message): state.phase = .failed(message)
            }

            return
        }

        state.entries = page.items
        state.nextCursor = page.nextCursor
        state.anomalies = loaded.1.value ?? []
        state.phase = page.items.isEmpty && state.anomalies.isEmpty ? .empty : .loaded
    }

    public func loadMore() async {
        guard let cursor = state.nextCursor, !state.loadingMore else { return }

        state.loadingMore = true
        defer { state.loadingMore = false }

        var filter = state.filter
        filter.cursor = cursor

        guard let page = try? await api.entries(filter) else { return }

        state.entries += page.items
        state.nextCursor = page.nextCursor
    }

    /// Changing a filter starts the list again: a cursor from the old filter
    /// would page through a different set of rows.
    public func filter(by action: AuditAction?) async {
        state.filter.action = action
        state.filter.cursor = nil
        state.nextCursor = nil
        state.phase = .loading

        await load()
    }
}

File: android/feature/audit/src/main/kotlin/xyz/klinik/feature/audit/AuditModel.kt
package xyz.klinik.feature.audit

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import xyz.klinik.network.ApiError
import xyz.klinik.network.AuditAction
import xyz.klinik.network.AuditAnomaly
import xyz.klinik.network.AuditApi
import xyz.klinik.network.AuditEntry
import xyz.klinik.network.AuditFilter
import xyz.klinik.network.UiText
import xyz.klinik.network.uiText

sealed interface AuditPhase {
    data object Loading : AuditPhase
    data object Loaded : AuditPhase

    /** Nothing matched the filter. Not a failure, and not an empty log. */
    data object Empty : AuditPhase

    /** Not a failure: reading the log needs `audit.read`. */
    data object NotPermitted : AuditPhase
    data class Failed(val message: UiText) : AuditPhase
}

data class AuditState(
    val phase: AuditPhase = AuditPhase.Loading,
    val entries: List<AuditEntry> = emptyList(),
    val anomalies: List<AuditAnomaly> = emptyList(),
    val action: AuditAction? = null,
    val nextCursor: String? = null,
    val error: UiText? = null,
) {
    val hasMore: Boolean get() = nextCursor != null
}

/**
 * Who did what to whose record (spec M13).
 *
 * The log is the thing that makes the rest of the clinic answerable, so this
 * model's whole job is to lose none of it: the page is appended rather than
 * replaced, nothing is filtered on the client, and an anomaly the server
 * flagged is carried in the server's own words.
 *
 * The anomalies are fetched beside the log rather than on their own screen.
 * They are the reason somebody opens this: "a nurse read a hundred and twenty
 * files last night" is not a row anybody would find by scrolling.
 */
class AuditModel(private val api: AuditApi) {
    private val _state = MutableStateFlow(AuditState())
    val state: StateFlow<AuditState> = _state.asStateFlow()

    suspend fun load() {
        _state.value = _state.value.copy(phase = AuditPhase.Loading)

        try {
            val page = api.entries(AuditFilter(action = _state.value.action))
            // Its own call, and allowed to fail on its own: a clinic that can
            // read the log should still get it when the detector is down.
            val anomalies = runCatching { api.anomalies() }.getOrDefault(emptyList())

            _state.value = _state.value.copy(
                phase = if (page.items.isEmpty()) AuditPhase.Empty else AuditPhase.Loaded,
                entries = page.items,
                anomalies = anomalies,
                nextCursor = page.nextCursor,
            )
        } catch (error: Throwable) {
            _state.value = _state.value.copy(
                phase = if (error is ApiError.Forbidden) {
                    AuditPhase.NotPermitted
                } else {
                    AuditPhase.Failed(messageFor(error))
                },
            )
        }
    }

    suspend fun choose(action: AuditAction?) {
        _state.value = _state.value.copy(action = action, entries = emptyList(), nextCursor = null)
        load()
    }

    suspend fun loadMore() {
        val cursor = _state.value.nextCursor ?: return

        val page = runCatching {
            api.entries(AuditFilter(action = _state.value.action, cursor = cursor))
        }.getOrNull() ?: return

        _state.value = _state.value.copy(
            entries = _state.value.entries + page.items,
            nextCursor = page.nextCursor,
        )
    }

    private fun messageFor(error: Throwable): UiText =
        (error as? ApiError)?.uiText() ?: UiText.Key("error.server")
}
